src/transformers/models/mamba2/modeling_mamba2.py

```diff
@@ -1801,34 +1801,80 @@ def forward(
             attentions=all_self_attns,
         )
 
-    def _update_causal_mask(self, attention_mask, inputs_embeds, past_key_values, output_attentions):
+    # Adapted from transformers.models.llama.modeling_llama.LlamaModel._update_causal_mask
+    # Custom hybrid cache instead
+    def _update_causal_mask(
+        self,
+        attention_mask: torch.Tensor,
+        input_tensor: torch.Tensor,
+        cache_position: torch.Tensor,
+        past_key_values: HybridMamba2AttentionDynamicCache,
+        output_attentions: bool,
+    ):
         if not self._uses_attention_layers:
             return None
 
-        batch_size, seq_len, _ = inputs_embeds.shape
-        past_length = past_key_values.get_seq_length() if past_key_values is not None else 0
-
-        # Follows GPTNeoX based creation of masks
-        attention_mask = attention_mask.view(batch_size, -1) if attention_mask is not None else None
         if self._attn_implementation == "flash_attention_2":
-            return attention_mask if (attention_mask is not None and 0 in attention_mask) else None
-        elif self._attn_implementation == "sdpa" and not output_attentions:
-            return _prepare_4d_causal_attention_mask_for_sdpa(
-                attention_mask=attention_mask,
-                input_shape=(batch_size, seq_len),
-                inputs_embeds=inputs_embeds,
-                past_key_values_length=past_length,
-            )
+            if attention_mask is not None and 0.0 in attention_mask:
+                return attention_mask
+            return None
+
+        # For SDPA, when possible, we will rely on its `is_causal` argument instead of its `attn_mask` argument, in
+        # order to dispatch on Flash Attention 2.
+        past_seen_tokens = past_key_values.get_seq_length() if past_key_values is not None else 0
+
+        # TODO: check if this is compatible with this custom cache format
+        if self._attn_implementation == "sdpa" and not output_attentions:
+            if AttentionMaskConverter._ignore_causal_mask_sdpa(
+                attention_mask,
+                inputs_embeds=input_tensor,
+                past_key_values_length=past_seen_tokens,
+                is_training=self.training,
+            ):
+                return None
```

Comment:
Not sure if this is correct since llama does not do it for a static cache. No idea why tho...